diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6675d2f..6769d8a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,6 +16,7 @@ ADD_LIBRARY(libfreecwmp_shared SHARED ${SOURCES})
 ADD_LIBRARY(libfreecwmp_static STATIC ${SOURCES})
 
 SET_TARGET_PROPERTIES(libfreecwmp_shared PROPERTIES OUTPUT_NAME freecwmp)
+SET_TARGET_PROPERTIES(libfreecwmp_shared PROPERTIES SOVERSION "0")
 SET_TARGET_PROPERTIES(libfreecwmp_static PROPERTIES OUTPUT_NAME freecwmp)
 
 INSTALL(TARGETS libfreecwmp_shared LIBRARY DESTINATION lib)
