diff --git a/meta-yocto/conf/bblayers.conf.sample b/meta-yocto/conf/bblayers.conf.sample
index 0eda565..82ec4d1 100644
--- a/meta-yocto/conf/bblayers.conf.sample
+++ b/meta-yocto/conf/bblayers.conf.sample
@@ -9,8 +9,17 @@ BBLAYERS ?= " \
   ##OEROOT##/meta \
   ##OEROOT##/meta-yocto \
   ##OEROOT##/meta-yocto-bsp \
+  ##OEROOT##/meta-raspberrypi \
+  ##OEROOT##/meta-openembedded/meta-oe \
+  ##OEROOT##/meta-openembedded/meta-python \
+  ##OEROOT##/meta-openembedded/meta-networking \
+  ##OEROOT##/meta-measured \
+  ##OEROOT##/iot-black/meta-sp \
+  ##OEROOT##/iot-black/meta-sp-raspberrypi \
   "
 BBLAYERS_NON_REMOVABLE ?= " \
   ##OEROOT##/meta \
   ##OEROOT##/meta-yocto \
   "
+
+BBMASK = "meta-raspberrypi/recipes-bsp/u-boot/u-boot-rpi_git.bb|meta-measured/networking-layer/recipes-support/strongswan/strongswan_5.3.0.bbappend"
