From 031836af7f5f7cb76c2aef18b35e5490c33cf39c Mon Sep 17 00:00:00 2001
From: KyungMo Kim <kkim@securityplatform.co.kr>
Date: Thu, 14 Jul 2016 16:17:43 +0900
Subject: [PATCH 03/16] firmware.inc: Update revision to include rpi3 support

---
 recipes-bsp/common/firmware.inc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/recipes-bsp/common/firmware.inc b/recipes-bsp/common/firmware.inc
index 020f0ed..2c4f497 100644
--- a/recipes-bsp/common/firmware.inc
+++ b/recipes-bsp/common/firmware.inc
@@ -1,5 +1,5 @@
-RPIFW_SRCREV ?= "f7108cfdc715f79c5fd051fb4c3ae6f4b1b01f23"
-RPIFW_DATE ?= "20151021"
+RPIFW_SRCREV ?= "028b1f6585280bd0a21885973589c4b101355597"
+RPIFW_DATE ?= "20160303"
 RPIFW_SRC_URI ?= "git://github.com/raspberrypi/firmware.git;protocol=git;branch=master"
 RPIFW_S ?= "${WORKDIR}/git"
 
-- 
1.9.1

