diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 49badba..f0eeb65 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -5,7 +5,9 @@ dtb-$(CONFIG_ARCH_BCM2708) += bcm2708-rpi-b-plus.dtb
 dtb-$(CONFIG_ARCH_BCM2708) += bcm2708-rpi-cm.dtb
 dtb-$(CONFIG_ARCH_BCM2835) += bcm2835-rpi-cm.dtb
 dtb-$(CONFIG_ARCH_BCM2709) += bcm2709-rpi-2-b.dtb
+dtb-$(CONFIG_ARCH_BCM2709) += bcm2709-rpi-2-b-tpm.dtb
 dtb-$(CONFIG_ARCH_BCM2709) += bcm2710-rpi-3-b.dtb
+dtb-$(CONFIG_ARCH_BCM2709) += bcm2710-rpi-3-b-tpm.dtb
 
 # Raspberry Pi
 ifeq ($(CONFIG_ARCH_BCM2708),y)
diff --git a/arch/arm/boot/dts/bcm2709-rpi-2-b-tpm.dts b/arch/arm/boot/dts/bcm2709-rpi-2-b-tpm.dts
new file mode 100644
index 0000000..351dd50
--- /dev/null
+++ b/arch/arm/boot/dts/bcm2709-rpi-2-b-tpm.dts
@@ -0,0 +1,15 @@
+#include "bcm2709-rpi-2-b.dts"
+
+&i2c1 {
+	status = "ok";
+	tpm_i2c_infineon {
+		type = "slb9645tt";
+		compatible = "infineon,slb9645tt";
+		reg = <0x20>;
+	};
+	tpm_i2c_atmel {
+		type = "tpm_i2c_atmel";
+		compatible = "atmel,at97sc3204t";
+		reg = <0x29>;
+	};
+};
diff --git a/arch/arm/boot/dts/bcm2710-rpi-3-b-tpm.dts b/arch/arm/boot/dts/bcm2710-rpi-3-b-tpm.dts
new file mode 100644
index 0000000..2efb10c
--- /dev/null
+++ b/arch/arm/boot/dts/bcm2710-rpi-3-b-tpm.dts
@@ -0,0 +1,15 @@
+#include "bcm2710-rpi-3-b.dts"
+
+&i2c1 {
+	status = "ok";
+	tpm_i2c_infineon {
+		type = "slb9645tt";
+		compatible = "infineon,slb9645tt";
+		reg = <0x20>;
+	};
+	tpm_i2c_atmel {
+		type = "tpm_i2c_atmel";
+		compatible = "atmel,at97sc3204t";
+		reg = <0x29>;
+	};
+};
